dbms/src/Flash/FlashService.cpp

```diff
@@ -623,55 +626,76 @@ grpc::Status FlashService::EstablishDisaggTask(grpc::ServerContext * grpc_contex
         current_memory_tracker = nullptr;
     });
 
-    grpc::Status ret_status = grpc::Status::OK;
-    auto record_error = [&](grpc::StatusCode err_code, int flash_err_code, const String & err_msg) {
+    auto record_error = [&](int flash_err_code, const String & err_msg) {
+        // Note: We intentinally do not remove the snapshot from the SnapshotManager
+        // when this request is failed. Consider this case:
+        // EstablishDisagg for A: ---------------- Failed --------------------------------------------- Cleanup Snapshot for A
+        // EstablishDisagg for B: - Failed - RN retry EstablishDisagg for A+B -- InsertSnapshot for A+B ----- FetchPages (Boom!)
         auto * err = response->mutable_error();
         err->set_code(flash_err_code);
         err->set_msg(err_msg);
-        ret_status = grpc::Status(err_code, err_msg);
     };
 
     try
     {
         handler->prepare(request);
         handler->execute(response);
     }
+    catch (const RegionException & e)
+    {
+        for (const auto & region_id : e.unavailable_region)
+        {
+            auto * retry_region = response->add_retry_regions();
+            retry_region->set_id(region_id);
+            // Note: retry_region's version and epoch is not set, because we miss these information
+            // from the exception.
+        }
+        record_error(
+            ErrorCodes::DISAGG_ESTABLISH_RETRYABLE_ERROR,
+            fmt::format("Retryable error: {}", e.message()));
+    }
+    catch (const LockException & e)
+    {
+        auto * retry_region = response->add_retry_regions();
+        retry_region->set_id(e.region_id);
+        // Note: retry_region's version and epoch is not set, because we miss these information
+        // from the exception.
+
+        // TODO: We may need to send this error back to TiDB? Otherwise TiDB
+        // may not resolve lock in time.
+        record_error(
+            ErrorCodes::DISAGG_ESTABLISH_RETRYABLE_ERROR,
+            fmt::format("Retryable error: {}", e.message()));
+    }
     catch (Exception & e)
     {
-        LOG_ERROR(logger, "EstablishDisaggTask meet Exception: {}\n{}", e.displayText(), e.getStackTrace().toString());
-        record_error(grpc::StatusCode::INTERNAL, e.code(), e.message());
+        LOG_WARNING(logger, "EstablishDisaggTask meet exception: {}\n{}", e.displayText(), e.getStackTrace().toString());
```

Comment:
these are non-retryable errors that make queries failed, why use warning instead of error level?